### PR TITLE
Build tests with warnings as errors enabled

### PR DESCRIPTION
**Pull request by t-b**
I'm trying to resurrect https://github.com/nlohmann/json/pull/1900.

- [x] Cleanup commits
- [x] `#if DOCTEST_CLANG >=` -> `#if DOCTEST_CLANG && DOCTEST_CLANG >=`
- [x] Fix https://github.com/nemequ/hedley/issues/40 upstream
- [x] Get it working

Close https://github.com/nlohmann/json/issues/1798.

### CODE REVIEW

**Comment by nlohmann** [comment]
FYI: These are the flags I check regularly: https://github.com/nlohmann/json/blob/develop/Makefile#L93.

File: include/nlohmann/thirdparty/hedley/hedley.hpp [hunk 1]
```diff
@@ -504,7 +504,8 @@
 #if \
     defined(__has_cpp_attribute) && \
     defined(__cplusplus) && \
-    (!defined(JSON_HEDLEY_SUNPRO_VERSION) || JSON_HEDLEY_SUNPRO_VERSION_CHECK(5,15,0))
+    (!defined(JSON_HEDLEY_SUNPRO_VERSION) || JSON_HEDLEY_SUNPRO_VERSION_CHECK(5,15,0)) && \
```
Comment (nlohmann) [inline include/nlohmann/thirdparty/hedley/hedley.hpp:507]: Why did you change the Hedley header? We copy this from https://github.com/nemequ/hedley, so the change is best made upstream.

File: test/src/unit-allocator.cpp [hunk 2]
```diff
@@ -96,6 +100,8 @@ struct my_allocator : std::allocator<T>
     {
         if (next_deallocate_fails)
         {
+            (void)p; // Ignored unused
```
Comment (gregmarr) [inline test/src/unit-allocator.cpp:103]: Why is this needed?  It's used in the `else` branch.

File: include/nlohmann/thirdparty/hedley/hedley.hpp (hunk 1 above)
Comment (t-b) [reply include/nlohmann/thirdparty/hedley/hedley.hpp:507]: @nlohmann This is only for intermediate hacking to get it working. I've already opened https://github.com/nemequ/hedley/issues/40.

File: test/src/unit-allocator.cpp (hunk 2 above)
Comment (t-b) [reply test/src/unit-allocator.cpp:103]: @gregmarr MSVC 2015 does warn here. I presume it first uses the fact that `next_deallocate_fails == true` ditches the else part and then complains about `p` not being used.

Comment (gregmarr) [reply test/src/unit-allocator.cpp:103]: That's very strange, because `next_deallocate_fails` is only ever `false`.  Ah well, fun with MSVC.

File: test/src/unit-items.cpp [hunk 3]
```diff
@@ -39,6 +39,18 @@ using nlohmann::json;
...
+    DOCTEST_CLANG_SUPPRESS_WARNING("-Wrange-loop-construct")
+#endif
+#endif
+
+#ifdef DOCTEST_PLATFORM_MAC
+#if DOCTEST_CLANG && DOCTEST_CLANG >= DOCTEST_COMPILER(12, 0, 0)
+    DOCTEST_CLANG_SUPPRESS_WARNING("-Wrange-loop-analysis")
+#endif
+#endif
+
```
Comment (t-b) [inline test/src/unit-items.cpp:53]: @nlohmann Any idea why the MacOSX part here does not work? https://travis-ci.org/github/nlohmann/json/jobs/707957810 still fails.

Comment (t-b) [reply test/src/unit-items.cpp:60]: This is due to https://github.com/onqtam/doctest/issues/415, I have a workaround.

**Comment by t-b** [comment]
@nlohmann Ready.

File: .github/workflows/windows.yml [hunk 4]
```diff
@@ -15,20 +15,6 @@ jobs:
     - name: test
       run: cd build ; ctest -j 10 -C Debug --exclude-regex "test-unicode" --output-on-failure
 
-  clang9:
```
Comment (nlohmann) [inline .github/workflows/windows.yml:18]: Why did you remove Clang 9?

File: test/src/unit-allocator.cpp [hunk 5]
```diff
@@ -29,6 +29,10 @@ SOFTWARE.
 
 #include "doctest_compatibility.h"
 
+#if DOCTEST_CLANG && DOCTEST_CLANG >= DOCTEST_COMPILER(3, 6, 0)
+    DOCTEST_CLANG_SUPPRESS_WARNING("-Wkeyword-macro")
```
Comment (nlohmann) [inline test/src/unit-allocator.cpp:33]: Do you still need this after `JSON_TESTS_PRIVATE` was introduced?

**Review by nlohmann — CHANGES_REQUESTED** [review]

File: .github/workflows/windows.yml (hunk 4 above)
Comment (t-b) [reply .github/workflows/windows.yml:18]: The commit message explains why, see e2264601542bd29a88fc85570b95c4e897eb2781.

File: test/src/unit-allocator.cpp (hunk 5 above)
Comment (t-b) [reply test/src/unit-allocator.cpp:33]: I'll have a look if I find time.

**Comment by MBalszun** [comment]
Is it possible to override this? 
E.g. when I want to compile the tests with a new compiler (with new warnings) on ci or my local computer.

**Comment by t-b** [comment]
@MBalszun Not as of now. But a patch is welcome to add a cmake option JSON_WARNINGS_AS_ERRORS defaulting to ON.

File: .github/workflows/windows.yml (hunk 4 above)
Comment (MBalszun) [reply .github/workflows/windows.yml:18]: Could the problem be that windows-latest doesn't remain the same? Newer versions of Visual studio's standard c++ library also require newer versions of clang, so when testing llvm on windows, a fixed version of clang should probably be paired with a fixed version of the VS toolchain.

**Comment by MBalszun** [comment]
So, what is the state here? Can this be merged? I see there is a merge conflict, but github doesn't show me what it is.

**Comment by nlohmann** [comment]
I am currently overworking the targets for the CI. The goal is to move everything from the individual Travis/AppVeyor/GitHub Actions files as well as the project's Makefile into some CMake-based targets. This includes the warning flags for GCC and Clang I linked in https://github.com/nlohmann/json/pull/2245#issuecomment-654457381. Then I think we only need to add a similar target for MSVC.

I will open a PR to make the proposal visible. It is not done yet, but I think it will tackle most issues described here (apart of MSVC).

**Comment by nlohmann** [comment]
See https://github.com/nlohmann/json/pull/2561.

**Comment by MBalszun** [comment]
Thanks for the update. I wasn't aware that all the warning flags for gcc and clang are set in a separate make file.

That aside: You are aware tha travis.org is being shut down and that travis.com won't provide unlimited CI time for open source projects anymore right? Or have you negotiated a deal with them?

**Comment by nlohmann** [comment]
I have no deal with travis. If they stop working the way they used to, I will rely on AppVeyor and GitHub Actions. In the meantime, I see whether I can get a self-hosted runner to work so that I can automate as many system-independent tests as possible.
